src/artm/regularizer_sandbox/smooth_sparse_phi.cc

```diff
@@ -53,55 +38,32 @@ bool SmoothSparsePhi::RegularizePhi(::artm::core::Regularizable* topic_model, do
     has_dictionary = false;
   }
 
-  if (!has_dictionary) {
   // proceed the regularization
-    for (int topic_id = 0; topic_id < topic_size; ++topic_id) {
-      for (int token_id = 0; token_id < topic_model->token_size(); ++token_id) {
-        if (topics_to_regularize[topic_id]) {
-          bool regularize_this_token = false;
-          if (!use_all_classes) {
-            if (std::find(classes_to_regularize.begin(),
-                          classes_to_regularize.end(),
-                          topic_model->token(token_id).class_id)
-                != classes_to_regularize.end()) {
-              regularize_this_token = true;
-            }
-          } else {
-            regularize_this_token = true;
-          }
-          if (regularize_this_token) {
-            topic_model->IncreaseRegularizerWeight(token_id, topic_id, static_cast<float>(tau));
-          }
-        }
+  for (int token_id = 0; token_id < topic_model->token_size(); ++token_id) {
+    float coefficient = 1.0f;
+    auto token = topic_model->token(token_id);
+    if (has_dictionary) {
+      if (use_all_classes ||
+          core::is_member(token.class_id, config_.class_id())) {
+        auto entry_iter = dictionary_ptr->find(token);
+        // don't process tokens without value in the dictionary
+        if (entry_iter == dictionary_ptr->end())
+          coefficient = 0.0f;
+        else
+          coefficient = entry_iter->second.value();
       }
     }
-  } else {
-    // proceed the regularization
+    bool use_this_token = core::is_member(token.class_id, config_.class_id());
```

Comment:
if (!use_all_classes && !use_this_token) continue;

In this case its nice to avoid "if" inside innermost loop. The code is equally readable, and second version is obviously more efficient. Also move "float value = ..." outside of the innermost loop since it is the same for all topics.